### PR TITLE
[RISCV][GISEL] Legalization, register bank selection, and instruction selection for scalable G_SELECT

### PR DESCRIPTION
**Pull request by michaelmaitland**
SelectionDAG has SELECT and VSELECT

    SELECT restricts the condition operand to an i1 and the true and false operands
    can be vectors. The result of a SELECT has the same type as the true and
    false operands.

    VSELECT has a vector condition operand and the true and false operands
    must be vectors. The result of a VSELECT has a vector result.
  
    GlobalISel has G_SELECT which has a condition operand that can be an
    i1 or can be an i1 vector. A G_SELECT acts like a ISD::SELECT when the
    condition operand is scalar, and an ISD::VSELECT when the condition is vector.

    In this patch, we would like to take advantage of the patterns written
    for SELECT and VSELECT, so we mark G_SELECT equivalent to both SELECT
    and VSELECT to reuse the patterns.

    We will probably need to test an i1 condition with vector true and false
    operands in the future. That can be the topic of another patch.

### CODE REVIEW

**Comment by michaelmaitland** [comment]
This PR depends on https://github.com/llvm/llvm-project/pull/84553 but I didn't include these in the commits of this PR to make it clear what commits need to be reviewed as part of this PR. I will rebase this PR on top of https://github.com/llvm/llvm-project/pull/84553 once it is merged. For now, the tests will expectedly fail.

**Comment by tschuett** [comment]
If I read the MachineVerifier correctly, then `G_SELECT (s1),
(vector-ty), (vector-ty)` is valid.

https://github.com/llvm/llvm-project/blob/c51095f51b770f0663b65dd8a651601fca15701e/llvm/lib/CodeGen/MachineVerifier.cpp#L1371

**Comment by michaelmaitland** [comment]
> If I read the MachineVerifier correctly, then `G_SELECT (s1), (vector-ty), (vector-ty)` is valid.
> 
> https://github.com/llvm/llvm-project/blob/c51095f51b770f0663b65dd8a651601fca15701e/llvm/lib/CodeGen/MachineVerifier.cpp#L1371

In that case, I made a mistake in my description. I wonder if we even need a separate opcode and we can just reuse G_SELECT as ISD::SELECT union ISD::VSELECT

**Comment by tschuett** [comment]
G_SELECT is defined as:
```
// Generic select
def G_SELECT : GenericInstruction {
  let OutOperandList = (outs type0:$dst);
  let InOperandList = (ins type1:$tst, type0:$src1, type0:$src2);
  let hasSideEffects = false;
}
```
It picks one of the two inputs. It is up to you to specify the relation between $type1 and $type0.

**Comment by arsenm** [comment]
> In that case, I made a mistake in my description. I wonder if we even need a separate opcode and we can just reuse G_SELECT as ISD::SELECT union ISD::VSELECT.
>

Yes. It never made sense to me why the DAG had a separate opcode, always seemed like pointless deviation from how IR works (same with fcmp/icmp->setcc)

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -202,8 +202,11 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
       .clampScalar(1, sXLen, sXLen)
       .clampScalar(0, sXLen, sXLen);
 
-  auto &SelectActions = getActionDefinitionsBuilder(G_SELECT).legalFor(
-      {{s32, sXLen}, {p0, sXLen}});
+  auto &SelectActions =
+      getActionDefinitionsBuilder(G_SELECT)
+          .legalFor({{s32, sXLen}, {p0, sXLen}})
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:207]: non-0 address space pointers?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:207]: Can we do that in a different patch? I am taking the existing legalFor from scalar and only adding for vectors. I specific in the PR title that this is just for scalable G_SELECT.

File: llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp [hunk 2]
```diff
@@ -401,6 +401,15 @@ RISCVRegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
   case TargetOpcode::G_SELECT: {
     LLT Ty = MRI.getType(MI.getOperand(0).getReg());
 
+    if (Ty.isVector()) {
```
Comment (tschuett) [inline llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:404]: You could try
```
 GSelect *Sel = cast<GSelect>(&MI);
```
, but ...

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:404]: Are you suggesting: `MI.getOperand(2).getReg()` -> `GSelect *Sel = cast<GSelect>(&MI); MRI.getType(Sel->getCondReg())`?

Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:404]: Exactly. Then you don't have to remember the operand indices and the code becomes more readable.

**Comment by topperc** [comment]
> and an ISD::VSELECT when the condition is scalar.

That seems wrong. An ISD::VSELECT has vector condition not scalar condition.

File: llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp [hunk 3]
```diff
@@ -401,6 +402,16 @@ RISCVRegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
...
+      GSelect *Sel = cast<GSelect>(&MI);
+      LLT TestTy = MRI.getType(Sel->getCondReg());
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:407]: assert that TestTy is vector

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:407]: Updated

**Comment by michaelmaitland** [comment]
> > and an ISD::VSELECT when the condition is scalar.
> 
> That seems wrong. An ISD::VSELECT has vector condition not scalar condition.

Revised

**Review by arsenm — APPROVED** [review]

**Comment by michaelmaitland** [comment]
Committed in 8c51ac9ddb026ca247b25b08b1d08e32b1ed58ec, 973e9dbd57bcdfcc3e0deef37d1c37fc4fbbb856, and 05840c8714d798c8f2873c205aeac146b4993ae5.

**Pull request closed without merging**